### DISCUSSION TITLE
Merging Datasets

### DISCUSSION THREAD

**Discussion by Ixiodor** [corteva/rioxarray#244] [Q&A] [answered]
I loaded N files(single bands) in a Dataset:
```
collected_ds = []
datasetname = xarray.Dataset()
datasetname['aname'] = rioxarray.open_rasterio(...)).squeeze() # Doing squeeze to resolve error about non 2D/3D array on to_raster
...
collected_ds.append(datasetname)
...
collected_ds.append(datasetname1)
```
and i did a clip on each dataset.
Now i collected N datasets that need to be geospatially merged for a mosaic:
`mosaic_test = rioxarray.merge.merge_arrays(collected_ds, precision=50, nodata=0)`
but that throws:

> rioxarray.exceptions.InvalidDimensionOrder: Invalid dimension order. Expected order: ('y', 'x')You can use `DataArray.transpose('y', 'x')` to reorder your dimensions. 

Why happens?

**Comment by snowman2** [comment, marked as answer]
Do you have to convert it to a dataset? If so, I recommend using:
```python
rioxarray.merge.merge_datasets(collected_ds, precision=50, nodata=0)
```

EDIT: Fix in #245 

Workaround: wait to do squeeze until after merging.

**Reply by Ixiodor** [reply]
They are altredy datasets, with N "names" to access their DataArrays.
Using merge_datasets throw:

> IndexError: too many indices for array: array is 2-dimensional, but 3 were indexed

**Reply by snowman2** [reply]
Are you able to show the `repr` of the datasets for more context? Also, the full traceback would be helpful.

**Reply by Ixiodor** [reply]
Here the full error:
> ../earthalytics_core/earthalytics_engine/Sentinel2l2a.py:281: in generate_images
>     mosaic_10m = rioxarray.merge.merge_datasets(products_10m, precision=50, nodata=0)
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rioxarray/merge.py:238: in merge_datasets
>     merged_data[data_var] = merge_arrays(
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rioxarray/merge.py:157: in merge_arrays
>     merged_data, merged_transform = _rio_merge(
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rasterio/merge.py:326: in merge
>     copyto(region, temp, region_nodata, temp_nodata,
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rasterio/merge.py:246: in copyto
>     old_data[mask] = new_data[mask]
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = masked_array(
>   data=[[nan, 291.0, 352.0, ..., nan, nan, nan],
>         [nan, 307.0, 376.0, ..., nan, nan, nan],
>       ... False, False, ..., False, False, False],
>         [False, False, False, ..., False, False, False]],
>   fill_value=1e+20)
> indx = array([[[ True,  True,  True, ...,  True,  True,  True],
>         [ True,  True,  True, ...,  True,  True,  True],
>     ...e],
>         [ True,  True,  True, ...,  True,  True,  True],
>         [ True,  True,  True, ...,  True,  True,  True]]])
> 
>     def __getitem__(self, indx):
>         """
>         x.__getitem__(y) <==> x[y]
>     
>         Return the item described by i, as a masked array.
>     
>         """
>         # We could directly use ndarray.__getitem__ on self.
>         # But then we would have to modify __array_finalize__ to prevent the
>         # mask of being reshaped if it hasn't been set up properly yet
>         # So it's easier to stick to the current version
> >       dout = self.data[indx]
> E       IndexError: too many indices for array: array is 2-dimensional, but 3 were indexed
> 

repr:

> print(repr(products_10m[0]))
> <xarray.Dataset>
> Dimensions:      (x: 249, y: 148)
> Coordinates:
>   * y            (y) float64 4.526e+06 4.526e+06 ... 4.525e+06 4.525e+06
>   * x            (x) float64 4.263e+05 4.264e+05 ... 4.288e+05 4.288e+05
>     band         int64 1
>     spatial_ref  int64 0
> Data variables:
>     blue         (y, x) float64 dask.array<chunksize=(148, 249), meta=np.ndarray>
>     green        (y, x) float64 dask.array<chunksize=(148, 249), meta=np.ndarray>
>     red          (y, x) float64 dask.array<chunksize=(148, 249), meta=np.ndarray>
>     nir          (y, x) float64 dask.array<chunksize=(148, 249), meta=np.ndarray>

**Reply by snowman2** [reply]
What happens if you don't do `squeeze()`?

**Reply by Ixiodor** [reply]
Seems to get merged but to_raster() throws:

> ../earthalytics_core/earthalytics_engine/Sentinel2l2a.py:284: in generate_images
>     mosaic_10m.rio.to_raster(m10_filepath)
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rioxarray/raster_dataset.py:410: in to_raster
>     return data_array.rio.to_raster(
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rioxarray/raster_array.py:915: in to_raster
>     count=int(self.count),
> /home/ubuntu/.cache/pypoetry/virtualenvs/earthalytics-core-YgAKQ8v--py3.8/lib/python3.8/site-packages/rioxarray/rioxarray.py:684: in count
>     extra_dim = self._check_dimensions()
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <rioxarray.raster_array.RasterArray object at 0x7efbcf155520>
> 
>     def _check_dimensions(self):
>         """
>         This function validates that the dimensions 2D/3D and
>         they are are in the proper order.
>     
>         Returns
>         -------
>         str or None: Name extra dimension.
>         """
>         extra_dims = list(set(list(self._obj.dims)) - set([self.x_dim, self.y_dim]))
>         if len(extra_dims) > 1:
> >           raise TooManyDimensions(
>                 "Only 2D and 3D data arrays supported."
>                 f"{_get_data_var_message(self._obj)}"
>             )
> E           rioxarray.exceptions.TooManyDimensions: Only 2D and 3D data arrays supported.

repr:

> <xarray.Dataset>
> Dimensions:      (band: 1, x: 249, y: 148)
> Coordinates:
>   * y            (y) float64 4.526e+06 4.526e+06 ... 4.525e+06 4.525e+06
>   * x            (x) float64 4.263e+05 4.264e+05 ... 4.288e+05 4.288e+05
>   * band         (band) int64 1
>     spatial_ref  int64 0
> Data variables:
>     blue         (band, y, x) float64 dask.array<chunksize=(1, 148, 249), meta=np.ndarray>
>     green        (band, y, x) float64 dask.array<chunksize=(1, 148, 249), meta=np.ndarray>
>     red          (band, y, x) float64 dask.array<chunksize=(1, 148, 249), meta=np.ndarray>
>     nir          (band, y, x) float64 dask.array<chunksize=(1, 148, 249), meta=np.ndarray>

**Reply by snowman2** [reply]
I think the fix here is to wait to do `squeeze()` until right before `to_raster()`.

**Reply by snowman2** [reply]
Fix in #245

**Reply by Ixiodor** [reply]
Worked, Thanks
